src/charm.py

```diff
@@ -262,6 +267,50 @@ def update_external_services(self) -> None:
 
         self.expose_external = self.model.config["expose-external"]
 
+    def update_tls_sans(self) -> None:
+        """Emits a certificate expiring event when sans in current certificates are out of date.
+
+        This can occur for a variety of reasons:
+        1. Node port has been toggled on
+        2. Node port has been toggled off
+        3. The public K8s IP has changed
+        """
+
+        for internal in [True, False]:
+            # if the certificate has already been requested, we do not want to re-request
+            # another one and lead to an infinite chain of certificate events.
+            if self.tls.is_set_waiting_for_cert_to_update(internal):
+                continue
+
+            current_sans = self.tls.get_current_sans(internal)
+            current_sans_ip = set(current_sans["sans_ips"]) if current_sans else set()
+            expected_sans_ip = (
+                set(self.tls.get_new_sans()["sans_ips"]) if current_sans else set()
+            )
+            sans_ip_changed = current_sans_ip ^ expected_sans_ip
+
+            if not sans_ip_changed:
+                continue
+
+            logger.info(
+                (
+                    f'Mongos {self.unit.name.split("/")[1]} updating certificate SANs - '
+                    f"OLD SANs = {current_sans_ip - expected_sans_ip}, "
+                    f"NEW SANs = {expected_sans_ip - current_sans_ip}"
+                )
+            )
+
+            self.tls.certs.on.certificate_expiring.emit(
+                certificate=self.tls.get_tls_secret(
+                    internal=internal, label_name=Config.TLS.SECRET_CERT_LABEL
+                ),
+                expiry=datetime.now().isoformat(),
+            )
+            # without this, it is possible that we constantly request new certificates
+            # over and over again - blocking the event queue from processing the expired
+            # certificate event.
+            self.tls.set_waiting_for_cert_to_update(waiting=True, internal=internal)
```

Comment:
This is a bit hacky. I prefer we simplify this part and offload this to the mongodb_tls module by exposing a method and calling it from here, e.g:
```
def request_new_certificates(self, internal: bool) -> None:
    csr = # fetch csr from secrets
    self.certs.request_certificate_revocation(csr)
    self.request_certificate(....)
```

Reply:
[the changes you request must be made as a separate pr](https://github.com/canonical/mongodb-operator/pull/504) - only once it is merged can I update this one